File: VB/XPCustomBindingExample/XPO/XpoHelper.vb
Imports System.Collections.Concurrent
Imports System.Collections.Generic
Imports System.Configuration
Imports DevExpress.Xpo
Imports DevExpress.Xpo.DB
Imports DevExpress.Xpo.Metadata

Namespace XPCustomBindingExample.XPO
	Public NotInheritable Class XpoHelper

		Private Sub New()
		End Sub

		Private Shared ConnectionProviders As New ConcurrentDictionary(Of String, IDataStore)()
		''' <summary>
		''' This option reduces unnecessary SELECT operations 
		''' Change this option if the inheritance mapping is required
		''' </summary>
		Public Shared DontLoadClassInfoFromDatabase As Boolean = True

		Shared Sub New()
			XpoDefault.Session = Nothing
			XpoDefault.DataLayer = Nothing
		End Sub

		Public Shared Function GetNewSession(ByVal connectionName As String, ByVal databaseSchema As XPDictionary) As Session
			Dim dataLayer As IDataLayer = GetDataLayer(connectionName, databaseSchema)
			Return New Session(dataLayer)
		End Function


		Public Shared Function GetNewUnitOfWork(ByVal connectionName As String, ByVal databaseSchema As XPDictionary) As UnitOfWork
			Dim dataLayer As IDataLayer = GetDataLayer(connectionName, databaseSchema)
			Return New UnitOfWork(dataLayer)
		End Function

		Private Shared Function GetDataLayer(ByVal connectionName As String, ByVal databaseSchema As XPDictionary) As IDataLayer
			Dim connectionProvider As IDataStore = GetConnectionProvider(connectionName)
			Dim dataLayer As IDataLayer = New SimpleDataLayer(databaseSchema, connectionProvider)

			If DontLoadClassInfoFromDatabase Then
				ConfigureDataLayerNotToLoadClassInfoFromDatabase(dataLayer)
			End If

			Return dataLayer
		End Function

		Private Shared Function GetConnectionProvider(ByVal connectionName As String) As IDataStore
			Return ConnectionProviders.GetOrAdd(connectionName, AddressOf GetConnectionPoolProvider)
		End Function

		Private Shared Function GetConnectionPoolProvider(ByVal connectionName As String) As IDataStore
			Dim connectionString As String = ConfigurationManager.ConnectionStrings(connectionName).ConnectionString
			connectionString = XpoDefault.GetConnectionPoolString(connectionString)
			Return XpoDefault.GetConnectionProvider(connectionString, AutoCreateOption.SchemaAlreadyExists)
		End Function

		Private Shared Sub ConfigureDataLayerNotToLoadClassInfoFromDatabase(ByVal dataLayer As IDataLayer)
			Dim objectLayer As New SimpleObjectLayer(dataLayer)
			Dim loadedTypes As New Dictionary(Of XPClassInfo, XPObjectType)()
			objectLayer.SetObjectLayerWideObjectTypes(loadedTypes)
		End Sub
	End Class
End Namespace

File: CS/XPCustomBindingExample/XPO/XpoHelper.cs
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Configuration;
using DevExpress.Xpo;
using DevExpress.Xpo.DB;
using DevExpress.Xpo.Metadata;

namespace XPCustomBindingExample.XPO {
    public static class XpoHelper {
        private static ConcurrentDictionary<string, IDataStore> ConnectionProviders = new ConcurrentDictionary<string,IDataStore>();
        /// <summary>
        /// This option reduces unnecessary SELECT operations 
        /// Change this option if the inheritance mapping is required
        /// </summary>
        public static bool DontLoadClassInfoFromDatabase = true;

        static XpoHelper() {
            XpoDefault.Session = null;
            XpoDefault.DataLayer = null;
        }

        public static Session GetNewSession(string connectionName, XPDictionary databaseSchema) {
            IDataLayer dataLayer = GetDataLayer(connectionName, databaseSchema);
            return new Session(dataLayer);
        }


        public static UnitOfWork GetNewUnitOfWork(string connectionName, XPDictionary databaseSchema) {
            IDataLayer dataLayer = GetDataLayer(connectionName, databaseSchema);
            return new UnitOfWork(dataLayer);
        }

        private static IDataLayer GetDataLayer(string connectionName, XPDictionary databaseSchema) {
            IDataStore connectionProvider = GetConnectionProvider(connectionName);
            IDataLayer dataLayer = new SimpleDataLayer(databaseSchema, connectionProvider);

            if(DontLoadClassInfoFromDatabase)
                ConfigureDataLayerNotToLoadClassInfoFromDatabase(dataLayer);

            return dataLayer;
        }

        private static IDataStore GetConnectionProvider(string connectionName) {
            return ConnectionProviders.GetOrAdd(connectionName, GetConnectionPoolProvider);
        }

        private static IDataStore GetConnectionPoolProvider(string connectionName) {
            string connectionString = ConfigurationManager.ConnectionStrings[connectionName].ConnectionString;
            connectionString = XpoDefault.GetConnectionPoolString(connectionString);
            return XpoDefault.GetConnectionProvider(connectionString, AutoCreateOption.SchemaAlreadyExists);
        }

        private static void ConfigureDataLayerNotToLoadClassInfoFromDatabase(IDataLayer dataLayer) {
            SimpleObjectLayer objectLayer = new SimpleObjectLayer(dataLayer);
            Dictionary<XPClassInfo, XPObjectType> loadedTypes = new Dictionary<XPClassInfo, XPObjectType>();
            objectLayer.SetObjectLayerWideObjectTypes(loadedTypes);
        }
    }
}